### PR TITLE
#84 Protect uuid from changes to invalid value

### PR DESCRIPTION
**Pull request by zerc**

### CODE REVIEW

File: pyuploadcare/api_resources.py [hunk 1]
```diff
@@ -22,6 +22,7 @@
 
 
 RE_UUID = '[a-z0-9]{8}-(?:[a-z0-9]{4}-){3}[a-z0-9]{12}'
+RE_UUID_REGEX = re.compile(RE_UUID, re.VERBOSE)
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: why verbose?

File: pyuploadcare/api_resources.py [hunk 2]
```diff
@@ -65,6 +66,20 @@ def __str__(self):
         return self.cdn_url
 
     @property
+    def uuid(self):
+        return self._uuid
+
+    @uuid.setter
+    def uuid(self, value):
+        match = RE_UUID_REGEX.match(value)
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: this will silently eat invalid values

File: pyuploadcare/api_resources.py [hunk 3]
```diff
@@ -65,6 +66,20 @@ def __str__(self):
...
+
+        if not match:
+            raise InvalidRequestError(
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: exception name is weird

File: tests/functional/test_api_resources.py [hunk 4]
```diff
@@ -19,8 +19,15 @@
 class FileRegexTest(unittest.TestCase):
 
     def test_value_error_when_uuid_is_bad(self):
-        file_serialized = 'blah'
-        self.assertRaises(InvalidRequestError, File, file_serialized)
+        file_serialized_invalid = 'blah'
+        self.assertRaises(InvalidRequestError, File, file_serialized_invalid)
+
+        file_serialized_valid = '3addab78-6368-4c55-ac08-22412b6a2a4c'
```
Comment (dmitry-mukhin) [inline tests/functional/test_api_resources.py]: also check '3addab78-6368-4c55-ac08-22412b6a2a4cblah'

File: pyuploadcare/api_resources.py (hunk 1 above)
Comment (zerc) [inline pyuploadcare/api_resources.py]: stupid copy paste :(
Fixed

File: pyuploadcare/api_resources.py (hunk 2 above)
Comment (zerc) [inline pyuploadcare/api_resources.py]: You right. I Did it more strict now. Also updated test case.

File: tests/functional/test_api_resources.py (hunk 4 above)
Comment (zerc) [inline tests/functional/test_api_resources.py]: Done.

File: pyuploadcare/api_resources.py [hunk 5]
```diff
@@ -65,6 +66,19 @@ def __str__(self):
...
     @property
+    def uuid(self):
+        return self._uuid
+
+    @uuid.setter
+    def uuid(self, value):
+        match = RE_UUID_REGEX.match(value)
+
+        if not match:
+            raise InvalidRequestError('Invalid UUID: {0}'.format(value))
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: so why InvalidRequestError ?

Comment (zerc) [inline pyuploadcare/api_resources.py]: Hm, you right, `InvalidRequestError` incorrect in this case but docstring says:

```python
class InvalidRequestError(UploadcareException, ValueError):
    """Invalid parameters errors, e.g. status 404."""
```

i.e. phrase `Invalid parameters errors` confusing me

And this exception also raises in `__init__` code where we just validate parameters without doing of request to service.

I propose to make a separate exception for all these cases e.g:

```python
class InvalidParamError(UploadcareException, ValueError):
   """Invalid parameters errors, e.g. invalid UUID"""

class InvalidRequestError(InvalidParam):
   """Invalid service parameters errors, e.g status 404"""
```

What do you think about this?

Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: okeyish
@homm ?

Comment (homm) [inline pyuploadcare/api_resources.py]: ```class InvalidParamError(UploadcareException, ValueError)```
:+1:

**Pull request merged**